File: Rust/rust_aoc/src/year_2015/d6_light_war.rs
use regex::Regex;
use rust_core::geometry::UCoord;

pub fn part_one(lines: &Vec<String>) -> String {
    let mut grid: Vec<bool> = vec![false; 1_000_000];

    lines.iter()
    .map(|l| parse_command(l))
    .for_each(|cmd| apply_one(&cmd, &mut grid));
    grid.into_iter().map(|v| if v { 1 } else { 0 })
    .sum::<u32>()
    .to_string()
}

pub fn part_two(lines: &Vec<String>) -> String {
    let mut grid: Vec<u32> = vec![0; 1_000_000];

    lines.iter()
    .map(|l| parse_command(l))
    .for_each(|cmd| apply_two(&cmd, &mut grid));

    grid.into_iter().sum::<u32>()
    .to_string()
}

enum CommandType {
    On,
    Off,
    Toggle
}

struct Command {
    bottom_left: UCoord,
    top_right: UCoord,
    command_type: CommandType
}

fn parse_command(line: &str) -> Command {
    let re = Regex::new(r"(toggle|turn on|turn off)\s(\d+,\d+)\sthrough\s(\d+,\d+)").unwrap();
    let Some(caps) = re.captures(line) else { panic!("Failed to match regex")};
    let bl = UCoord::from_str(&caps[2]);
    let tr = UCoord::from_str(&caps[3]);

    return match &caps[1] {
        "toggle" => Command{bottom_left: bl, top_right: tr, command_type: CommandType::Toggle},
        "turn on" => Command{bottom_left: bl, top_right: tr, command_type: CommandType::On},
        "turn off" => Command{bottom_left: bl, top_right: tr, command_type: CommandType::Off},
        _ => panic!("Unrecognized command type")
    };
}

fn apply_one(cmd: &Command, grid: &mut Vec<bool>) {
    for x in cmd.bottom_left.x .. cmd.top_right.x + 1 {
        for y in cmd.bottom_left.y .. cmd.top_right.y + 1 {
            match cmd.command_type {
                CommandType::On => grid[x + y * 1000usize] = true,
                CommandType::Off => grid[x + y * 1000usize] = false,
                _ => grid[x + y * 1000usize] = !grid[x+ y * 1000usize]
            }
        }
    }

}

fn apply_two(cmd: &Command, grid: &mut Vec<u32>) {
    for x in cmd.bottom_left.x .. cmd.top_right.x + 1 {
        for y in cmd.bottom_left.y .. cmd.top_right.y + 1 {
            let light = &mut grid[x + y * 1000usize];

            match cmd.command_type {
                CommandType::On => *light += 1,
                CommandType::Off => if *light > 0 { *light -= 1; },
                CommandType::Toggle => *light += 2
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_command_tests() {
        let c1 = parse_command("turn on 0,0 through 999,999");
        match c1.command_type {
            CommandType::On => (),
            _ => assert!(false)
        }
        assert!(c1.bottom_left == UCoord::new(0, 0));
        assert!(c1.top_right == UCoord::new(999, 999));

        let c2 = parse_command("toggle 0,0 through 999,0");
        match c2.command_type {
            CommandType::Toggle => (),
            _ => assert!(false)
        }
        assert!(c2.bottom_left == UCoord::new(0,0));
        assert!(c2.top_right == UCoord::new(999,0));

        let c3 = parse_command("turn off 499,499 through 500,500");
        match c3.command_type {
            CommandType::Off => (),
            _ => assert!(false)
        }
        assert!(c3.bottom_left == UCoord::new(499,499));
        assert!(c3.top_right == UCoord::new(500,500));
    }

    #[test]
    fn part_one_tests() {
        let lines = vec!["turn on 0,0 through 1,1", "toggle 1,1 through 2,2"].iter().map(|s| String::from(*s)).collect();
        assert_eq!(part_one(&lines), "6");
    }
}

File: Cpp/AdventOfCode/src/2015/d6_LightWar.cpp
#pragma once

#include "2015/d6_LightWar.h"

SOLUTION(2015, 6) {
    enum struct CommandType { On, Off, Toggle };

    struct Command {
        UCoord BottomLeft{};
        UCoord TopRight{};
        CommandType CommandType{};
    };

    constexpr Command ParseCommand(std::string_view line) {
        auto s1 = Constexpr::Split(line, " through ");
        auto s2 = Constexpr::Split(s1[0], " ");
        auto blStr = Constexpr::Split(s2.back(), ",");
        auto trStr = Constexpr::Split(s1[1], " ");

        Command result;
        result.CommandType = s2.size() == 2 ? CommandType::Toggle :
            s2[1] == "on" ? CommandType::On : CommandType::Off;

        result.BottomLeft = UCoord(s2.back());
        result.TopRight = UCoord(s1[1]);

        return result;
    }

    constexpr std::vector<size_t> GetLightIndexes(const Command& command, size_t gridSize) {
        std::vector<size_t> result;
        for (auto x = command.BottomLeft.X; x <= command.TopRight.X; x++) {
            for (auto y = command.BottomLeft.Y; y <= command.TopRight.Y; y++) {
                result.push_back(x * gridSize + y);
            }
        }
        return result;
    }

    constexpr size_t SolvePartOne(auto& lines) {
        constexpr size_t gridSize = 1000;
        auto gridPtr = new std::array<bool, gridSize* gridSize>();
        auto grid = *gridPtr;

        for (const auto& line : lines) {
            auto command = ParseCommand(line);
            auto indexes = GetLightIndexes(command, gridSize);
            switch (command.CommandType) {
            case CommandType::On: for (auto i : indexes) grid[i] = true; break;
            case CommandType::Off: for (auto i : indexes) grid[i] = false; break;
            case CommandType::Toggle: for (auto i : indexes) grid[i] = !grid[i]; break;
            }
        }

        auto count = std::count(grid.cbegin(), grid.cend(), true);
        delete gridPtr;

        return count;
    }
    PART_ONE() {
        return Constexpr::ToString(SolvePartOne(lines));
    }

    constexpr size_t SolvePartTwo(auto& lines) {
        constexpr size_t gridSize = 1000;
        size_t* grid = new size_t[gridSize * gridSize]();

        for (const auto& line : lines) {
            auto command = ParseCommand(line);
            auto indexes = GetLightIndexes(command, gridSize);
            switch (command.CommandType) {
            case CommandType::On: for (auto i : indexes) grid[i]++; break;
            case CommandType::Off: for (auto i : indexes) grid[i] = grid[i] == 0 ? 0 : grid[i] - 1; break;
            case CommandType::Toggle: for (auto i : indexes) grid[i] += 2; break;
            }
        }

        size_t result = 0;
        for (size_t i = 0; i < gridSize * gridSize; i++) {
            result += grid[i];
        }

        delete[] grid;
        return result;
    }

    PART_TWO() {
        return Constexpr::ToString(SolvePartTwo(lines));
    }

    constexpr bool TestParseCommand() {
        auto c1 = ParseCommand("turn on 0,0 through 999,999");
        if (c1.CommandType != CommandType::On) return false;
        if (c1.BottomLeft != UCoord{ 0, 0 }) return false;
        if (c1.TopRight != UCoord{ 999, 999 }) return false;

        auto c2 = ParseCommand("toggle 0,0 through 999,0");
        if (c2.CommandType != CommandType::Toggle) return false;
        if (c2.BottomLeft != UCoord{ 0, 0 }) return false;
        if (c2.TopRight != UCoord{ 999,0 }) return false;

        auto c3 = ParseCommand("turn off 499,499 through 500,500");
        if (c3.CommandType != CommandType::Off) return false;
        if (c3.BottomLeft != UCoord{ 499, 499 }) return false;
        if (c3.TopRight != UCoord{ 500,500 }) return false;

        return true;
    }

    TESTS() {
        if (!TestParseCommand()) return false;

        std::vector<std::string> lines = {
            "turn on 0,0 through 1,1",
            "toggle 1,1 through 2,2"
        };
        if (SolvePartOne(lines) != 6) return false;

        return true;
    }
}